### PR TITLE
Store finalized blockHashes 

### LINKED ISSUE #3252: Cache fault tolerance in block metdata
## Overview
Looking for fault tolerance for old blocks is resource consuming. There should be a way to avoid unnecessary computation.

### Impact
<https://github.com/rchain/rchain/pull/3251> and <https://github.com/rchain/rchain/pull/2947> should be reverted

### Design
it is reasonable to accumulate and cache fault tolerance in block metadata

### PR DESCRIPTION
**Pull request by zsluedem**
## Overview
Store finalized blockHashes 

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by nzpr** [comment]
<https://github.com/rchain/rchain/pull/3251> this should be reverted as well

**Comment by zsluedem** [comment]
> #3251 this should be reverted as well

It should be overwritten in this pr.

**Comment by nzpr** [comment]
I believe it's better to store fault tolerances, as isFinalized just checks if fault tolerance is greater than threshold. And threshold potentially can be dynamic. Not sure though whats the best way to implement this

File: casper/src/test/scala/coop/rchain/casper/engine/Setup.scala [hunk 1]
```diff
@@ -133,8 +136,12 @@ object Setup {
           estimateBlockHash: BlockHash
       ): Task[Float] = Task.pure(1.0f)
     }
-    implicit val lastFinalizedBlockCalculator = LastFinalizedBlockCalculator[Task](0f)
-    implicit val synchronyConstraintChecker   = SynchronyConstraintChecker[Task](0d)
+    val finalizedBlocksStore  = InMemoryKeyValueStore[Task]()
+    val finalizedBlocksStoreT = finalizedBlocksStore.toTypedStore(codecBlockHash, codecBlockHash)
```
Comment (nzpr) [inline casper/src/test/scala/coop/rchain/casper/engine/Setup.scala]: As you need to store only list of hashes, you can store only keys, so for value use codec `ignore(0)`, it is defined in scodec package. And use `KeyValueTypedStore[F, BlockHash, Unit]`

**Comment by nzpr** [comment]
I have been thinking, and dynamic threshold, if present, should not matter - if shard decide to rise the threhold for finalization, it should not influence previous finalized block. In that case to finalize child bigger fault tolerance will be required.

File: casper/src/test/scala/coop/rchain/casper/engine/Setup.scala (hunk 1 above)
Comment (zsluedem) [reply casper/src/test/scala/coop/rchain/casper/engine/Setup.scala]: Nice idea

**Comment by zsluedem** [comment]
close because of https://github.com/rchain/rchain/pull/3258

**Pull request closed without merging**